interface.py

```diff
@@ -572,51 +574,53 @@ def upload():
 
     files = request.files.getlist("file[]")
     if not files:
-        return redirect("./", code=400)
+        abort(400)
 
-    # filename = secure_filename(file.filename).strip()
     for file in files:
         filename = file.filename
         if filename == '':
-            return redirect("./", code=400)
+            abort(400)
 
         targetdir = request.form['targetdir'].strip()
         if targetdir == '':
             targetdir = 'uploads/'
         elif '../' in targetdir:
-            return redirect("./", code=400)
+            abort(403)
+
+        filename = secure_filename(file.filename).strip()
 
-        log.info('web: Uploading file from %s:' % request.remote_addr)
-        log.info('web: - filename: ' + filename)
-        log.info('web: - targetdir: ' + targetdir)
-        log.info('web: - mimetype: ' + file.mimetype)
+        log.info(f'web: Uploading file from {request.remote_addr}:')
+        log.info(f'web: - file name: {filename}')
+        log.info(f'web: - target dir: {targetdir}')
+        log.info(f'web: - mime type: {file.mimetype}')
 
         if "audio" in file.mimetype:
             storagepath = os.path.abspath(os.path.join(var.music_folder, targetdir))
-            print('storagepath:', storagepath)
             if not storagepath.startswith(os.path.abspath(var.music_folder)):
-                return redirect("./", code=400)
+                abort(403)
 
             try:
                 os.makedirs(storagepath)
             except OSError as ee:
                 if ee.errno != errno.EEXIST:
-                    return redirect("./", code=500)
+                    log.error(f'web: failed to create directory {storagepath}')
+                    abort(500)
 
             filepath = os.path.join(storagepath, filename)
-            log.info(' - filepath: ' + filepath)
+            log.info(f'web: - save path: {filepath}')
             if os.path.exists(filepath):
                 continue
 
             file.save(filepath)
         else:
-            continue
+            log.error(f'web: unsupported file type {file.mimetype}! File was not saved.')
+            abort(415)
```

Comment:
This is an interesting interaction. 

I might be misunderstanding how the upload process goes - if one abort stops all future and past uploads of this request, then don't mind what I'm writing below.

I think it's correct to return `415` in this case, but this particular way of triggering it can be prey to weird scenarios for multi-file upload. Imagine this: I'm uploading two files. One of a supported format and one of an unsupported format. But in what order am I uploading them?

So say it's a tuple of files `(a.mp3, b.alac)`. `alac` is not supported, so it will be rejected. Here's how it will go down:

- `a.mp3` is uploaded
- `b.alac` is rejected
- Request returns `415` but `a.mp3` was still uploaded.

Whereas if the order is inverted:

- `b.alac` is rejected
- `a.mp3` is never checked and also rejected
- Request returns `415` and nothing was uploaded (as its number would suggest!).

If I understood this code correctly, the idea is to reject everything if one of the items is unsupported, right?
Then maybe we could do a check over all items in the list, and reject them immediately if one of them is unsupported. And after that, moving on to iterate over the list to upload it :)